File: app/src/lib/receipt-pdf.ts
import { jsPDF } from "jspdf";
import { save } from "@tauri-apps/plugin-dialog";
import { invoke } from "@tauri-apps/api/core";
import { formatDateTime } from "@/lib/utils";

export interface ReceiptData {
  saleId: number;
  model: string;
  imei: string;
  date: string;
  price: number;
  paymentLabel: string;
  commissionAmount: number;
  warrantyText: string | null;
  buyerName: string | null;
}

const COMPANY_NAME = "NEVA MOBILE";

const TR_TO_ASCII: Record<string, string> = {
  ş: "s", Ş: "S", ı: "i", İ: "I", ğ: "g", Ğ: "G",
  ü: "u", Ü: "U", ö: "o", Ö: "O", ç: "c", Ç: "C",
};

/**
 * jsPDF'nin standart (Helvetica) fontu Türkçe özel karakterleri (ş,ı,ğ,ü,ö,ç)
 * ve ₺ işaretini WinAnsi kod sayfasında bulamadığı için bozuk gösteriyor
 * (ör. "Satış" -> "Sat1_"). Unicode font gömmek yerine (ek bağımlılık, risk)
 * PDF'e özel ASCII'ye çeviriyoruz — okunabilir ve garantili doğru render olur.
 */
function trAscii(s: string): string {
  return s.replace(/[şŞıİğĞüÜöÖçÇ]/g, (c) => TR_TO_ASCII[c] ?? c);
}

/** PDF'te ₺ sembolü yerine "TL" son eki — Helvetica'da ₺ glifi yok. */
function formatKurusPlain(kurus: number): string {
  const lira = kurus / 100;
  const text = lira.toLocaleString("tr-TR", { maximumFractionDigits: 2 });
  return `${trAscii(text)} TL`;
}

/**
 * Satış fişini PDF olarak üretir ve kullanıcının seçtiği konuma yazar.
 * jsPDF'nin kendi `.save()` çağrısı tarayıcı indirme API'sine (Blob/`<a download>`)
 * dayanır ve WebView2 içinde çalışmaz; bunun yerine PDF bayt olarak üretilip
 * Tauri kayıt diyaloğu + `write_binary_file` komutuyla gerçek dosyaya yazılır.
 * Kullanıcı diyaloğu iptal ederse sessizce çıkar.
 */
export async function generateReceiptPdf(data: ReceiptData): Promise<void> {
  const doc = new jsPDF({ unit: "mm", format: "a5" });
  const pageWidth = doc.internal.pageSize.getWidth();
  const marginX = 14;
  let y = 18;

  // Başlık
  doc.setFont("helvetica", "bold");
  doc.setFontSize(18);
  doc.setTextColor(37, 51, 219);
  doc.text(COMPANY_NAME, pageWidth / 2, y, { align: "center" });
  y += 6;
  doc.setFont("helvetica", "normal");
  doc.setFontSize(10);
  doc.setTextColor(110, 110, 120);
  doc.text(trAscii("Telefon Alım Satım Yönetim Sistemi"), pageWidth / 2, y, { align: "center" });
  y += 4;
  doc.setDrawColor(230, 230, 236);
  doc.line(marginX, y, pageWidth - marginX, y);
  y += 10;

  doc.setFontSize(14);
  doc.setTextColor(20, 20, 26);
  doc.setFont("helvetica", "bold");
  doc.text(trAscii("Satış Fişi"), marginX, y);
  doc.setFont("helvetica", "normal");
  doc.setFontSize(9);
  doc.setTextColor(140, 140, 150);
  doc.text(trAscii(`Fiş No: #${data.saleId}`), pageWidth - marginX, y, { align: "right" });
  y += 10;

  const rows: [string, string][] = [
    [trAscii("Telefon Modeli"), trAscii(data.model)],
    ["IMEI", data.imei],
    [trAscii("Satış Tarihi"), formatDateTime(data.date)],
    [trAscii("Alıcı Adı"), trAscii(data.buyerName ?? "-")],
    [trAscii("Ödeme Türü"), trAscii(data.paymentLabel)],
    [trAscii("Satış Fiyatı"), formatKurusPlain(data.price)],
    ...(data.commissionAmount > 0
      ? ([[trAscii("Banka Komisyonu"), `-${formatKurusPlain(data.commissionAmount)}`]] as [
          string,
          string,
        ][])
      : []),
    ["Garanti", trAscii(data.warrantyText ?? "Yok")],
  ];

  doc.setFontSize(10.5);
  for (const [label, value] of rows) {
    doc.setTextColor(120, 120, 130);
    doc.text(label, marginX, y);
    doc.setTextColor(20, 20, 26);
    doc.setFont("helvetica", "bold");
    doc.text(value, pageWidth - marginX, y, { align: "right" });
    doc.setFont("helvetica", "normal");
    y += 7.5;
  }

  y += 4;
  doc.setDrawColor(230, 230, 236);
  doc.line(marginX, y, pageWidth - marginX, y);
  y += 8;

  const netTotal = data.price - data.commissionAmount;
  doc.setFont("helvetica", "bold");
  doc.setFontSize(12);
  doc.setTextColor(37, 51, 219);
  doc.text(trAscii("Kasaya Giren"), marginX, y);
  doc.text(formatKurusPlain(netTotal), pageWidth - marginX, y, { align: "right" });

  y += 16;
  doc.setFont("helvetica", "normal");
  doc.setFontSize(8.5);
  doc.setTextColor(150, 150, 158);
  doc.text(trAscii("Bu fiş NEVA MOBILE tarafından otomatik olarak oluşturulmuştur."), pageWidth / 2, y, {
    align: "center",
  });

  const targetPath = await save({
    title: "PDF Satış Fişi",
    defaultPath: `Satis_Fisi_${data.saleId}.pdf`,
    filters: [{ name: "PDF", extensions: ["pdf"] }],
  });
  if (!targetPath) return;

  const bytes = Array.from(new Uint8Array(doc.output("arraybuffer")));
  await invoke("write_binary_file", { targetPath, bytes });
}

File: mobile/lib/core/pdf/receipt_pdf.dart
import 'dart:typed_data';

import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart' as pw;

import 'pdf_text.dart';

class ReceiptData {
  final int saleId;
  final String model;
  final String? imei;
  final String date;
  final int price;
  final String paymentLabel;
  final int commissionAmount;
  final String? warrantyText;
  final String? buyerName;

  ReceiptData({
    required this.saleId,
    required this.model,
    this.imei,
    required this.date,
    required this.price,
    required this.paymentLabel,
    this.commissionAmount = 0,
    this.warrantyText,
    this.buyerName,
  });
}

const _blue = PdfColor.fromInt(0xFF2533DB);
const _grayLabel = PdfColor.fromInt(0xFF78787E);
const _nearBlack = PdfColor.fromInt(0xFF14141A);
const _footerGray = PdfColor.fromInt(0xFF96969E);
const _divider = PdfColor.fromInt(0xFFE6E6EC);

/// Masaüstü receipt-pdf.ts ile aynı A5 fiş düzeni.
Future<Uint8List> buildReceiptPdf(ReceiptData data) async {
  final doc = pw.Document();
  final netTotal = data.price - data.commissionAmount;

  final rows = <List<String>>[
    [trAscii('Telefon Modeli'), trAscii(data.model)],
    [trAscii('IMEI'), data.imei ?? '-'],
    [trAscii('Satış Tarihi'), trAscii(data.date)],
    [trAscii('Alıcı Adı'), data.buyerName != null ? trAscii(data.buyerName!) : '-'],
    [trAscii('Ödeme Türü'), trAscii(data.paymentLabel)],
    [trAscii('Satış Fiyatı'), formatKurusPlain(data.price)],
    if (data.commissionAmount > 0)
      [trAscii('Banka Komisyonu'), '-${formatKurusPlain(data.commissionAmount)}'],
    [trAscii('Garanti'), data.warrantyText != null ? trAscii(data.warrantyText!) : 'Yok'],
  ];

  doc.addPage(
    pw.Page(
      pageFormat: PdfPageFormat.a5,
      margin: const pw.EdgeInsets.all(24),
      build: (context) {
        return pw.Column(
          crossAxisAlignment: pw.CrossAxisAlignment.start,
          children: [
            pw.Center(
              child: pw.Text('NEVA MOBILE',
                  style: pw.TextStyle(fontSize: 18, fontWeight: pw.FontWeight.bold, color: _blue)),
            ),
            pw.Center(
              child: pw.Text(trAscii('Telefon Alım Satım Yönetim Sistemi'),
                  style: const pw.TextStyle(fontSize: 10, color: _grayLabel)),
            ),
            pw.SizedBox(height: 10),
            pw.Divider(color: _divider),
            pw.SizedBox(height: 6),
            pw.Row(
              mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
              children: [
                pw.Text(trAscii('Satış Fişi'),
                    style: pw.TextStyle(fontSize: 14, fontWeight: pw.FontWeight.bold, color: _nearBlack)),
                pw.Text('Fis No: #${data.saleId}', style: const pw.TextStyle(fontSize: 9, color: _grayLabel)),
              ],
            ),
            pw.SizedBox(height: 8),
            for (final row in rows)
              pw.Padding(
                padding: const pw.EdgeInsets.symmetric(vertical: 3.5),
                child: pw.Row(
                  mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
                  children: [
                    pw.Text(row[0], style: const pw.TextStyle(fontSize: 10.5, color: _grayLabel)),
                    pw.Text(row[1],
                        style: pw.TextStyle(
                            fontSize: 10.5, fontWeight: pw.FontWeight.bold, color: _nearBlack)),
                  ],
                ),
              ),
            pw.SizedBox(height: 6),
            pw.Divider(color: _divider),
            pw.SizedBox(height: 6),
            pw.Row(
              mainAxisAlignment: pw.MainAxisAlignment.spaceBetween,
              children: [
                pw.Text(trAscii('Kasaya Giren'),
                    style: pw.TextStyle(fontSize: 12, fontWeight: pw.FontWeight.bold, color: _blue)),
                pw.Text(formatKurusPlain(netTotal),
                    style: pw.TextStyle(fontSize: 12, fontWeight: pw.FontWeight.bold, color: _blue)),
              ],
            ),
            pw.Spacer(),
            pw.Center(
              child: pw.Text(
                trAscii('Bu fiş NEVA MOBILE tarafından otomatik olarak oluşturulmuştur.'),
                style: const pw.TextStyle(fontSize: 8.5, color: _footerGray),
              ),
            ),
          ],
        );
      },
    ),
  );

  return doc.save();
}

String receiptFileName(int saleId) => 'Satis_Fisi_$saleId.pdf';
